src/persist-client/src/internal/state.rs

```diff
@@ -938,15 +929,46 @@ where
         self.writers.clear();
         self.leased_readers.clear();
         self.critical_readers.clear();
-        let mut new_trace = Trace::default();
-        new_trace.downgrade_since(&Antichain::new());
-        let merge_reqs = new_trace.push_batch(Self::tombstone_batch());
-        assert_eq!(merge_reqs, Vec::new());
-        self.trace = new_trace;
 
         debug_assert!(self.is_tombstone());
         Continue(())
     }
+
+    pub fn shrink_tombstone(&mut self) -> ControlFlow<NoOpStateTransition<()>, ()> {
+        let mut to_replace = None;
+        let mut batch_count = 0;
+        self.trace.map_batches(|b| {
+            batch_count += 1;
+            if !b.parts.is_empty() && to_replace.is_none() {
+                to_replace = Some(b.desc.clone());
+            }
+        });
+        if let Some(desc) = to_replace {
```

Comment:
Feels like this could use some commentary laying out how we expect this process to happen. Seems like we replace each batch one-by-one with an empty one and then once they're all empty, replace them all with a single batch?

(Btw, this is much simpler than what I had been picturing :clap:)

Reply:
Yeah, absolutely, good call.

Reply:
(And glad to hear! It's certainly a whole lot simpler than what I had cooking before your `apply_merge_res` suggestion...)